### PR TITLE
Decoupling the StableX contract and driver

### PR DESCRIPTION
**Pull request by fleupold**
This PR adds a layer of abstraction between the StableX driver and StableX contract. In particular, it separates the potential interactions in reading (OrderBookReading trait) and writing (SolutionSubmitting trait).

This allows us to implement more sophisticated logic for the two (e.g. an order book reader based on the graph instead of the contracts, a paginated reader, a reader that is using a blacklist of users or tokens, etc.) and decouples the driver from the smart contract.

In the next step I will implement a FilterinOrderBookReader that reads a blacklist from an environment variable.

Most of the added lines is boilerplate code. In particular I will look into replacing our mocking framework into something less verbose.

### Test Plan

Unit test.

### CODE REVIEW

File: driver/src/orderbook/mod.rs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+use crate::error::DriverError;
+
+use dfusion_core::models::{AccountState, Order};
+
+use web3::types::U256;
+
+type Result<T> = std::result::Result<T, DriverError>;
+
+pub trait StableXOrderBookReading {
+    /// Retunrs the index of the auction that is currently being solved
```
Comment (bh2smith) [inline driver/src/orderbook/mod.rs]:
```suggestion
    /// Returns the index of the auction that is currently being solved
```

File: driver/src/orderbook/mod.rs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    /// or an error in case it cannot get this information.
+    fn get_auction_index(&self) -> Result<U256>;
+
+    /// Returns the current state of the order book, inclducing account balances
```
Comment (bh2smith) [inline driver/src/orderbook/mod.rs]:
```suggestion
    /// Returns the current state of the order book, including account balances
```

File: driver/src/solution_submission/mod.rs [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+type Result<T> = std::result::Result<T, DriverError>;
+
+pub trait StableXSolutionSubmitting {
+    /// Return the objective value for the given solution in the given
+    /// batch or an error.
+    ///
+    /// # Arguments
+    /// - `batch_index` - the auction for which this solutions should be evaluated
+    /// - `orders` - the list of orders for which this solution is applicable
+    /// - `solution` - the solution to be evaluated
```
Comment (bh2smith) [inline driver/src/solution_submission/mod.rs]: Is this a JS doc string adaptation?

File: driver/src/orderbook/mod.rs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    /// and open orders or an error in case it cannot get this information.
+    ///
+    /// # Arguments
+    /// * `index` - the auction index for which returned orders should be valid
```
Comment (nlordell) [inline driver/src/orderbook/mod.rs]: ubernit: the notation is a bit inconsistent, `* name - description` here and `- name - description` in other places.

File: driver/src/solution_submission/mod.rs [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+impl<'a> StableXSolutionSubmitting for StableXSolutionSubmitter<'a> {
+    fn get_solution_objective_value(
+        &self,
+        batch_index: U256,
+        orders: Vec<Order>,
+        solution: Solution,
+    ) -> Result<U256> {
+        self.contract
+            .get_solution_objective_value(batch_index, orders, solution)
```
Comment (nlordell) [inline driver/src/solution_submission/mod.rs]: Maybe the `get_solution_objective_value` implementation should move here, and call the contract's `submit_solution` directly?

File: driver/src/solution_submission/mod.rs [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+    fn submit_solution(
+        &self,
+        batch_index: U256,
+        orders: Vec<Order>,
+        solution: Solution,
+        claimed_objective_value: U256,
+    ) -> Result<()> {
+        self.contract
+            .submit_solution(batch_index, orders, solution, claimed_objective_value)
```
Comment (nlordell) [inline driver/src/solution_submission/mod.rs]: Same here.

File: driver/src/solution_submission/mod.rs (hunk 5 above)
Comment (fleupold) [reply driver/src/solution_submission/mod.rs]: Yes, for this PR I wanted to keep the factoring as is. Theoretically we could also minify the trait and just expose the submit solution method and have the fact that it's first verified and then sent hidden to the driver.

File: driver/src/solution_submission/mod.rs (hunk 3 above)
Comment (fleupold) [reply driver/src/solution_submission/mod.rs]: I tried to follow https://doc.rust-lang.org/rust-by-example/meta/doc.html

File: driver/src/solution_submission/mod.rs (hunk 5 above)
Comment (nlordell) [reply driver/src/solution_submission/mod.rs]: Sounds like a great plan

**Comment by bh2smith** [comment]
Maybe merge master into here to fetch the changes you fix with the wait statement earlier today! It appear the build is never gonna finish.

**Comment by bh2smith** [comment]
How did you trigger build without changing any files?!

**Comment by fleupold** [comment]
> How did you trigger build without changing any files?!

`git commit --allow-empty`

**Pull request merged**